Zend/zend_operators.c

```diff
@@ -2274,69 +2274,55 @@ ZEND_API int ZEND_FASTCALL is_smaller_or_equal_function(zval *result, zval *op1,
 }
 /* }}} */
 
-static zend_bool ZEND_FASTCALL instanceof_interface_only(const zend_class_entry *instance_ce, const zend_class_entry *ce) /* {{{ */
+ZEND_API zend_bool ZEND_FASTCALL instanceof_class(const zend_class_entry *instance_ce, const zend_class_entry *ce) /* {{{ */
 {
-	uint32_t i;
-
-	for (i = 0; i < instance_ce->num_interfaces; i++) {
-		if (instanceof_interface_only(instance_ce->interfaces[i], ce)) {
+	do {
+		if (instance_ce == ce) {
 			return 1;
 		}
-	}
+	} while (instance_ce = instance_ce->parent);
+
 	return 0;
 }
 /* }}} */
 
-static zend_always_inline zend_bool instanceof_class(const zend_class_entry *instance_ce, const zend_class_entry *ce) /* {{{ */
+ZEND_API zend_bool ZEND_FASTCALL instanceof_interface(const zend_class_entry *instance_ce, const zend_class_entry *ce) /* {{{ */
 {
-	while (instance_ce) {
-		if (instance_ce == ce) {
+	uint32_t i;
+
+	/* The simplest case, an exact match */
+	if (instance_ce == ce) {
+		return 1;
+	}
+
+	/* Normally a ce will be directly aware of all the interfaces it inherits */
+	for (i = 0; i < instance_ce->num_interfaces; i++) {
+		if (instance_ce->interfaces[i] == ce) {
 			return 1;
 		}
-		instance_ce = instance_ce->parent;
 	}
-	return 0;
-}
-/* }}} */
 
-static zend_bool ZEND_FASTCALL instanceof_interface(const zend_class_entry *instance_ce, const zend_class_entry *ce) /* {{{ */
-{
-	uint32_t i;
+	/* Manual manipulation of the hierarchy is sometimes done by , but the changes
```

Comment:
Missing the most important part ... done by what?

Reply:
...done by several things runkit, uopz and doubtless many other weird and wonderful things cooked up by people who will come crawling out of the woodwork when their inadvisable magic that relies on undefined behaviour stops working.

I totally understand that there would be a preference for efficiency over supporting undefined behaviour, however I *think* I have come up with a way to have both, and imho we should at least *try* to avoid breakage where practical.

I'm closing this PR for now while I work on what I hope will be a solution that satisfies everyone, however if that does not prove to be the case then we can just go with the original approach. We should be able to move forward on this by the end of the week.